### PR TITLE
[Control] add labelElement prop

### LINKED ISSUE #1351: Allow ReactNode / component for IControlProps label
## Feature Request

Currently, `IControlProps.label` is typed as a `string`, but it should be a `ReactNode`. This should just be a change of typings. (Casting as `any` works 😁 )

For my specific use case, I wanted to have the label of a `Radio` component include a `code` element:
```
const labelHack = <span>You have selected <code>2</code></span> as any;
return <Radio label={labelHack} value="2" />;
```

Some other reasonable use cases:
- `<span><span className="pt-icon-standard pt-icon-person" /> Dr. Blue Print</span>`
- `<span>Label B <span className="pt-text-muted">(required)</span></span>` (From BP's [Form Groups example](http://blueprintjs.com/docs/#core/components/forms/form-group))

### PR DESCRIPTION
**Pull request by llorca**
#### Fixes #1351

A new prop must be added for TypeScript support because the existing `label` prop in `React.HTMLAttributes` only accepts `string` values, so tsc rejects JSX elements. 

JS consumers can happily pass JSX to the `label` prop (cuz no compiler to stop them and the code does actually support JSX) but we need to add a new prop to sidestep the typings for TS consumers.

### CODE REVIEW

**Comment by giladgray** [comment]
❌ THIS DOES NOT COMPILE. Try actually using it in JSX: 
```jsx
<Checkbox label={<strong>"Gilad Gray"</strong>} defaultIndeterminate />
```
and you'll see an error:
```
file: 'file:///Volumes/git/blueprint/packages/core/examples/controlsExample.tsx'
severity: 'Error'
message: 'Type 'Element' is not assignable to type 'string | (string & number) | (string & true) | (string & false) | (string & {}) | (string & React...'.
  Type 'Element' is not assignable to type 'string & (string | number | boolean | any[] | ReactElement<any>)[]'.
    Type 'Element' is not assignable to type 'string'.'
at: '26,27'
source: 'ts'
```

this is because `HTMLAttributes` defines `label: string | number` which _does not include `JSX.Element`_. you cannot simply add it to the types because the prop is defined twice, so all types must be compatible.

@llorca please add `labelElement` instead, as we discussed previously.

File: packages/core/src/components/forms/controls.tsx [hunk 1]
```diff
@@ -17,8 +17,11 @@ import * as React from "react";
 import * as Classes from "../../common/classes";
 import { IProps, removeNonHTMLProps } from "../../common/props";
 import { safeInvoke } from "../../common/utils";
+import { HTMLInputProps } from "../../index";
+
+export interface IControlProps extends IProps, HTMLInputProps {
```
Comment (llorca) [inline packages/core/src/components/forms/controls.tsx]: 🎉

File: packages/core/src/components/forms/controls.tsx [hunk 2]
```diff
@@ -31,22 +34,36 @@ export interface IControlProps extends IProps {
...
     // we spread props to `<input>` but render `children` as its sibling
     "children",
     "defaultIndeterminate",
     "indeterminate",
+    "labelElement",
 ];
 
 /** Base Component class for all Controls */
-export class Control<P extends IControlProps> extends React.Component<React.HTMLProps<HTMLInputElement> & P, {}> {
+export class Control<P extends IControlProps> extends React.Component<P, {}> {
```
Comment (llorca) [inline packages/core/src/components/forms/controls.tsx]: Do we still need `<P extends IControlProps>` if we already have interfaces extending `IControlProps`, such as `export interface ICheckboxProps extends IControlProps`?

File: packages/core/src/components/forms/controls.tsx [hunk 3]
```diff
@@ -62,6 +79,7 @@ export class Control<P extends IControlProps> extends React.Component<React.HTML
                 <input {...inputProps} ref={inputRef} type={type} />
                 <span className={Classes.CONTROL_INDICATOR} />
                 {this.props.label}
+                {this.props.labelElement}
```
Comment (llorca) [inline packages/core/src/components/forms/controls.tsx]: If both `label` and `labelElement` props are provided, do we still render both?

File: packages/core/src/components/forms/controls.tsx (hunk 2 above)
Comment (giladgray) [inline packages/core/src/components/forms/controls.tsx]: yes it enforces that the `P` is a subclass of `IControlProps`, so I can't pass `IDialogProps`, for instance. these two `extends` are separate things.

File: packages/core/src/components/forms/controls.tsx (hunk 3 above)
Comment (giladgray) [inline packages/core/src/components/forms/controls.tsx]: what does the code say?

yes, and `label` comes first.

Comment (llorca) [inline packages/core/src/components/forms/controls.tsx]: @giladgray Apologies, I meant to ask if we actually want to render both. Sounds error-prone to allow both at the same time

Comment (giladgray) [inline packages/core/src/components/forms/controls.tsx]: why? it's just JSX children, there's no limit. seems more error-prone to only render one.

Comment (llorca) [inline packages/core/src/components/forms/controls.tsx]: I just can't imagine a case where setting those two props at the same time would be intentional. Probably a false concern 🤷‍♂️

Comment (giladgray) [inline packages/core/src/components/forms/controls.tsx]: nor can i, but the simple thing works reasonably well in all cases here 🍰

File: packages/core/examples/controlExamples.tsx [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+// tslint:disable-next-line:max-classes-per-file
+export class SwitchExample extends BaseExample<IRadioExampleState> {
+    protected renderExample() {
+        return (
+            <div>
+                <label className={Classes.LABEL}>Privacy setting</label>
+                <Switch labelElement={<strong>Enabled</strong>} />
+                <Switch labelElement={<em>Public</em>} />
+                <Switch labelElement={<u>Cooperative</u>} defaultChecked />
+                <small>This example uses <code>labelElement</code> to demonstrate JSX labels.</small>
```
Comment (cmslewis) [inline packages/core/examples/controlExamples.tsx]: 👏

**Comment by thasner** [comment]
curious, why add `labelElement` instead of extending the typings for `label`?

**Comment by llorca** [comment]
@thasner scroll up and read the convo, especially this comment: https://github.com/palantir/blueprint/pull/1378#issuecomment-317844784

**Comment by thasner** [comment]
👍

**Pull request merged**